File: Programmers/[60060] 가사 검색/Johoseong/60060.py
def insert(trie, w, l):
    cur_node = trie
    for a in w:
        cur_node.setdefault(a, [{}, {}])
        if l not in cur_node[a][0]: # 자식 단어 길이 저장!!
            cur_node[a][0][l] = 0
        cur_node[a][0][l] += 1
        cur_node = cur_node[a][1]

def search(trie, q, l):
    cur_node = trie
    cnt = 0
    for a in q:
        if a == '?': # 물음표면? 지금까지 매칭된 단어들 길이랑 쿼리 길이 같은거 count
            for i in cur_node:
                if l in cur_node[i][0]:
                    cnt += cur_node[i][0][l]
            return cnt

        if a in cur_node: # 해당 알파벳이 트리에 있으면 다음노드로
            cur_node = cur_node[a][1]
        else:
            break
    return cnt

def solution(words, queries):
    answer = []
    trie = {}
    reverse_trie = {}

    for w in words:
        insert(trie, w, len(w)) # 일반적인 경우
        insert(reverse_trie, reversed(w), len(w)) # 단어 뒤집은 경우

    for q in queries:
        if q[0] == '?' and q[-1] == '?': # 다 물음표
            answer.append(search(trie, q, len(q)))
        elif q[0] == '?': # 앞에가 물음표 -> 뒤집어서 단어검색!!!!!!
            answer.append(search(reverse_trie, reversed(q), len(q)))
        else: # 뒤에가 물음표
            answer.append(search(trie, q, len(q)))
    return answer

File: Programmers/[60060] 가사 검색/woong-jae/60060.js
class Node {
    constructor() {
        this.count = 0;
        this.child = new Map();
    }
}

class Trie {
    constructor() {
        this.root = new Node();
    }
    insert(word) {
        let cur = this.root;
        if(!cur.child.has(word.length)) cur.child.set(word.length, new Node());
        cur = cur.child.get(word.length);
        
        for(let char of word) {
            if(!cur.child.has(char)) cur.child.set(char, new Node());
            cur.count++;
            cur = cur.child.get(char);
        }
    }
    getMatch(query) {
        let cur = this.root;
        if(!cur.child.has(query.length)) return 0;
        cur = cur.child.get(query.length);
        
        for(let char of query) {
            if(char === "?") return cur.count;
            if(!cur.child.has(char)) return 0;
            cur = cur.child.get(char);
        }
    }
}

function solution(words, queries) {
    const answer = [];
    
    const reverse = word => {
        return word.split("").reverse().join("");
    }
    
    const trie = new Trie();
    const reverseTrie = new Trie();
    
    words.forEach(word => {
        trie.insert(word);
        reverseTrie.insert(reverse(word));
    });
    
    queries.forEach(query => {
        answer.push(
            query[0] === "?" 
            ? reverseTrie.getMatch(reverse(query)) 
            : trie.getMatch(query)
        )
    });
    
    return answer;
}